### ISSUE TITLE
Run time and memory requirements for different imputation strategies

### ISSUE THREAD

**Issue by jaclyn-taroni** [AlexsLemonade/refinebio#751]
### Context

Imputation strategies for #508

### Problem or idea

We don't want processing the species compendia to take 20 years and we need a better idea of the run time and memory requirements for different imputation strategies that are determined to perform at an acceptable error rate. @Miserlou has already looked into this a little bit for [`KNNImputer`](https://libsleipnir.bitbucket.io/KNNImputer.html) and the [`impute`](https://www.bioconductor.org/packages/release/bioc/html/impute.html) R package. However, if we need to impute missing values in "batches" (of samples) for technical reasons, we should get a better idea of how many species might require such treatment and what the "missingness" of those species look like.  (The "missingness" data structure is beyond the scope of this ticket.)

### Solution or next step

We can use simulated data ("copy and paste" expansion of real masked data) to explore run time and memory requirements from different combinations of 1) number of genes and 2) number of samples.

#### Number of samples

Based on the species I expect we will have compendia for (e.g., have more than 100 samples on a supported Affymetrix platform), the range of potential sample sizes is ~500-801,000 (human). So we can pad this such that we're looking for values up to 1 million.

Here's how I would do the linear interpolation in R for a range of 50 sample sizes that I think mirrors the true distribution of the number of samples reasonably well :
```R
sample_size <- c(500, 10000, 50000, 1000000)
interpolated_values <- approx(sample_size, n = 50)$y
sample_grid <- ceiling(interpolated_values)
```

`sample_grid` then would be the different number of samples for testing. (Doesn't necessarily need to be 50.)

#### Number of genes 

Using the tsv files from the [identifier refinery](https://github.com/AlexsLemonade/identifier-refinery), we can get an idea of the range of the number of genes that will be included: ~3000-47000. Note that this is microarray only, which is very likely to be a lower bound compared to the RNA-seq data. So, we may want to check out 3K to 80K genes.

#### Methods

_To start_ -- the [`impute`](https://www.bioconductor.org/packages/release/bioc/html/impute.html) R package (Unless this has already been tried in some detail? I have not looked at "correctness" with this package yet), from [`fancyimpute`](https://pypi.org/project/fancyimpute/): `KNN`, `IterativeSVD`, `SoftImpute`

#### Instance types

@Miserlou does it make sense to try this on different instance types in your opinion?

Whoever tackles this, please contact me for data to "copy/paste" + we can coordinate such that we can do the "correctness"/error for masked values experiments in parallel.

### New Issue Checklist

- [x] The title is short and descriptive.
- [x] You have explained the context that led you to write this issue.
- [x] You have reported a problem or idea.
- [x] You have proposed a solution or next step.

**Comment by Miserlou** [comment]
To clarify, I've only looked at this for Sleipnir and the `impute` bioconductor package. Both seemed to perform fine with small datasets (at least in terms of viability, not correctness, which I didn't confirm) but weren't able to successfully compute with large datasets (100,000) - even when performed on an x32. ([Upstream issue here](https://bitbucket.org/libsleipnir/sleipnir/issues/28/segmentation-fault-with-100-000-samples), but I'm not holding my breath).

I'm worried about the runtime complexity of some of these algorithms for the size of our datasets. Both of the packages only ran a single core at a time, so it'd be great if we could find an implementation that parallelized better (assuming that the expected runtime is in the hundreds-of-days range, that could bring us down to the tens-of-days-range. If it's not going to take that long then this isn't as much of a priority - although finding one that we can modify to add a progress bar too would help!)

I haven't yet looked at `fancyimpute`, I'll take a stab at that today.

**Comment by Miserlou** [comment]
![screen shot 2018-10-22 at 4 12 47 pm](https://user-images.githubusercontent.com/139987/47316834-0cf7ef80-d616-11e8-9671-d2f1425c9331.png)

Early experiments with `fancyimpute` are promising, but it's slow as heck. Fortunately, I think it's "a few days" slow, not "a few years" slow.

**Comment by jaclyn-taroni** [comment]
We've found that `IterativeSVD`, which 
> Should be similar to SVDimpute from Missing value estimation methods for DNA microarrays by Troyanskaya et. al.

according to the [fancyimpute Github](https://github.com/iskandr/fancyimpute) to be considerably faster/appears to take advantage of parallelization (related: #754) compared to the fancyimpute implementation of KNN. All methods seem to have a similar _enough_ performance in preliminary experiments to suggest that this run time advantage should be heavily weighted.
